File: app/src/main/java/com/star_zero/example/androidmvvm/presentation/tasks/adapter/TasksAdapter.java
package com.star_zero.example.androidmvvm.presentation.tasks.adapter;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.star_zero.example.androidmvvm.R;
import com.star_zero.example.androidmvvm.databinding.ItemTaskBinding;
import com.star_zero.example.androidmvvm.domain.task.Task;

import java.util.List;

public class TasksAdapter extends RecyclerView.Adapter<TasksAdapter.ViewHolder> {

    private List<Task> tasks;

    @Override
    public TasksAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        LayoutInflater inflater = LayoutInflater.from(context);
        View view = inflater.inflate(R.layout.item_task, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(TasksAdapter.ViewHolder holder, int position) {
        holder.setTask(tasks.get(position));
    }

    @Override
    public int getItemCount() {
        if (tasks == null) return 0;
        return tasks.size();
    }

    public void setTasks(List<Task> tasks) {
        this.tasks = tasks;
        notifyDataSetChanged();
    }

    static class ViewHolder extends RecyclerView.ViewHolder {

        private ItemTaskBinding binding;

        private ItemTaskViewModel viewModel;

        ViewHolder(View itemView) {
            super(itemView);

            binding = DataBindingUtil.bind(itemView);

            viewModel = new ItemTaskViewModel();
            binding.setViewModel(viewModel);
        }

        void setTask(Task task) {
            viewModel.setTask(task);
            binding.executePendingBindings();
        }
    }
}

File: app/src/main/kotlin/com/star_zero/example/androidmvvm/presentation/tasks/adapter/TasksAdapter.kt
package com.star_zero.example.androidmvvm.presentation.tasks.adapter

import android.databinding.DataBindingUtil
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.star_zero.example.androidmvvm.R
import com.star_zero.example.androidmvvm.databinding.ItemTaskBinding
import com.star_zero.example.androidmvvm.domain.task.Task

class TasksAdapter : RecyclerView.Adapter<TasksAdapter.ViewHolder>() {

    private var tasks: List<Task>? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TasksAdapter.ViewHolder {
        val context = parent.context
        val inflater = LayoutInflater.from(context)
        val view = inflater.inflate(R.layout.item_task, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: TasksAdapter.ViewHolder, position: Int) {
        val tasks = this.tasks ?: return
        holder.setTask(tasks[position])
    }

    override fun getItemCount(): Int {
        return tasks?.size ?: 0
    }

    fun setTasks(tasks: List<Task>) {
        this.tasks = tasks
        notifyDataSetChanged()
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        private val binding: ItemTaskBinding = DataBindingUtil.bind<ItemTaskBinding>(itemView)

        private val viewModel: ItemTaskViewModel = ItemTaskViewModel()

        init {
            binding.viewModel = viewModel
        }

        fun setTask(task: Task) {
            viewModel.task = task
            binding.executePendingBindings()
        }
    }
}
